rfcs/010-gateway-refactor/010-gateway-refactor.md

```diff
@@ -0,0 +1,205 @@
+- one service could replace both traefik and the gateway
+- problems with managing sessions across different services/clients
+- using golang should result in more efficient use of resources and improved performance
+
+Reasons for centralized session authentication:
+- running sessions in subdomains which would allow for SSH access
+
+## Design Detail
+
+### Gateway
+
+The gateway has to have sessions with different clients. Some of these are:
+- User servers (i.e. the sidecar container needs be able to swap some kind of ID
+or access token for a Gitlab token)
+- Actual users logging in through the UI
+- Actual users logging in through the CLI
+
+Using JWT access tokens for this purpose is not feasible. This is because access
+tokens from Keycloak expire usually within an hour or several hours. This means
+that anyone who has to remain authenticated for a long time with the gateway
+needs to also get a refresh token and worry about using this to have valid access
+tokens. Accepting an expired access token is not secure.
+
+The gateway should instead have server-side sessions with every "client". For each such
+session the gateway will issue a random, hard to guess ID that will be stored either
+in a cookie or simply as a secret. This is exactly how sessions are handled by the
+ui-server. The cookie or ID is presented with every request to the gateway. The gateway
+looks this ID up in redis and if there is a match then any tokens/credentials stored
+under that key can be added to requests initiated by that ID.
+
+These sessions are managed/started as follows:
+- when a user logs in they get a session
+- when a user server is started the server gets a redis-based session
+- when a login happens through the CLI there is a also a session
+- anonymous users also have redis-based sessions but these are empty
+since there is no additional credentials that need to be managed for anonymous users
+
+The gateway then has the following responsibilities:
+- store refresh and access tokens
+- keep all tokens stored up to date and manage refreshing tokens
+- destroy sessions when certain events occur (i.e. when a logout occurs)
+
+### Centralized Jupyter server authentication
+
+### Diagrams
+
+1. Creating a user server
+
+```mermaid
+graph TD
+    Ingress -- 1<br>Session ID --> Gateway
+    
+    subgraph GatewayGroup[ ]
+    Gateway
+    DB[(DB)]
+    Gateway -- 2<br>Session ID --> DB
+    Gateway -- 3<br>Server ID --> DB
+    DB -- 4<br>KC access token<br>Server ID --> Gateway
+    end    
+    
+    Gateway -- 5<br>KC access token<br>Server ID--> Notebook[Notebook service]
+    Notebook -- 6<br>KC user ID<br>Server ID--> Server[User server]
+```
+
+2. Accessing a user server
+
+```mermaid
+graph TD
+    Ingress -- 1<br>Session ID --> Gateway
+    
+    subgraph GatewayGroup[ ]
+    Gateway
+    DB[(DB)]
+    Gateway -- 2<br>Session ID --> DB
+    DB -- 3<br>KC access token --> Gateway
+    end    
+    
+    Gateway -- 4<br>KC access token --> Proxy
```

Comment:
so there still is a proxy inside the session pod? That processes the JWT? The intro says there's no more oauth2 proxy in the sessions pod so I'm confused.

Reply:
@rokroskar I need to correct the intro. We need something either in the gateway or in the jupyter server to decide whether the request coming into to the jupyter server is from the right user. IMO this is an authorization question - not authentication, as such it makes sense to not make this the responsibility of the gateway. In this situation the gateway simply authenticates the user and injects the user's access token in the request. Then the jupyter server proxy can decide if that user should be granted access to the session or not.

If this is moved in the gateway the gateway would need to somehow figure out which user is the owner of the jupyter server being accessed. This means that we need a database with this info or the gateway would need to query k8s. I think that for this reason and as I mentioned above this should not be the responsibility of the gateway. When we implement a central authorization service this can be offloaded there. But for now it makes the most sense for this to remain in the jupyter server. I think this can be fulfilled by a container in the jupyter server that either has traefik with a simple config or some kind of configurable proxy. It is possible that the oauth2 proxy can be configured to handle this. But we can see and decide what is easiest and makes most sense.

Reply:
ok that's fine and that's the way I understood it from the diagrams, but please make it consistent (I realize this is evolving so no worries :)

Reply:
Done. The intro states this more clearly now.